### PR TITLE
feat: add generic pagination

### PR DESCRIPTION
**Pull request by kshitij-k-osmosys**
## Description

- Added paging request (for page number and page size) to be passed as query parameters
- Added a generic paging function that returns a PagedList<T> Item. It contains
  - `List<T>` result
  - `pagenumber`
  - `pagesize`
  - `totalcount`
  - `hasPrevious`
  - `hasNext`
- PagedList takes IQueryable, params to form a list<T> that can be manipulated further
- By default if no params are passed, pageNumber=1 and pageSize=10
- Created integration method `PaginationServices` with generic typing to be used in any repo
- Fixed routes of task

## Documentation
https://soochna.osmosys.co/books/research-development/page/generic-pagination

## Screenshots
request
request

pagination
Pagination response for Pagenumber = 2 and Pagesize = 3

## Summary by CodeRabbit

- **New Features**
	- Improved pagination for better performance and user experience.
	- Enhanced error handling and validation for task details.
- **Enhancements**
	- Updated API routes for task operations.
	- Improved exception handling in task services.
- **Documentation**
	- Enhanced inline documentation for pagination and task management.
- **Refactor**
	- Refactored task operations to utilize pagination.
- **Dependencies**
	- Added `Microsoft.EntityFrameworkCore` version `8.0.2` for enhanced database operations.

### CODE REVIEW

File: DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    public int PageSize
+    {
+        get
+        {
+            return pageSize;
+        }
+        set
+        {
+            pageSize = value > MaxPageSize ? MaxPageSize : value;
+        }
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: Fixed in next commit

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs [hunk 2]
```diff
@@ -20,104 +20,180 @@ public TaskController(ITaskDetailsService TaskDetailsService)
...
-    public async Task<ActionResult<BaseResponse<List<TaskDetailsResponse>>>> GetAllTasksAsync()
+    public async Task<ActionResult<BaseResponse<PagedList<TaskDetailsResponse>>>> GetAllTasksAsync([FromQuery] PagingRequest pagingRequest)
     {
-        BaseResponse<List<TaskDetailsResponse>> response = new(ResponseStatus.Fail);
-
-        response.Data = await _taskDetailsService.GetAllTasksAsync().ConfigureAwait(false);
-        response.Status = ResponseStatus.Success;
-
-        return Ok(response);
+        BaseResponse<PagedList<TaskDetailsResponse>> response = new(ResponseStatus.Fail);
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs:28]: What is the meaning on `new(ResponseStatus.Fail)`. It should be the instance of BaseResponse.

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs [hunk 3]
```diff
@@ -20,104 +20,180 @@ public TaskController(ITaskDetailsService TaskDetailsService)
...
+        }
     }
 
     /// <summary>
     /// Update details of a task when the Id is passed.
     /// </summary>
     /// <param name="taskId">Id of task record</param>
     /// <param name="modifiedDetails">Modified details for task record</param>
-    [HttpPut("{taskId}")]    
+    [HttpPut("tasks/{taskId}")]
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: This `task` URL should be singular.

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs [hunk 4]
```diff
@@ -20,104 +20,180 @@ public TaskController(ITaskDetailsService TaskDetailsService)
...
     [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
     [ProducesResponseType(StatusCodes.Status500InternalServerError)]
     public async Task<ActionResult<BaseResponse<TaskDetailsResponse>>> UpdateTaskAsync(int taskId, TaskDetailsRequest modifiedDetails)
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: `modifiedDetails` is not a meaningful variable name. The variable has to be related to task.

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs [hunk 5]
```diff
@@ -20,104 +20,180 @@ public TaskController(ITaskDetailsService TaskDetailsService)
...
+        try
+        {
+            response.Data = await _taskDetailsService.UpdateTaskAsync(taskId, modifiedDetails).ConfigureAwait(false);
+            response.Status = ResponseStatus.Success;
+
+            return Ok(response);
+        }
+        catch (NotFoundException ex)
+        {
+            response.Message = ex.Message;
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: @Harish-osmosys has taken care of this. Do you need to do it as well?

File: DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿namespace DotnetFoundation.Application.Models.Common;
+public class PagingRequest
+{
+    const int MaxPageSize = 50;
+    public int PageNumber { get; set; }
+    private int pageSize;
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: Why does it have to be in camel casing? Is there a standard that we are following?

File: DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿namespace DotnetFoundation.Application.Models.Common;
+public class PagingRequest
+{
+    const int MaxPageSize = 50;
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: Is this a public property? Access spec is not given here.

**Review by sameer-s-b — CHANGES_REQUESTED** [review]
Left a review and Suggested a few changes.

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs (hunk 5 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: I worked on this as I had better knowledge which error need to be handled

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs (hunk 2 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs:28]: It is creating a BaseResponse<T> Object `response` with ResponseStatus of enum Fail
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0090

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs (hunk 4 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: Renamed this to `updatedTaskDetails`

File: DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs (hunk 3 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Api/Controllers/TaskController.cs]: Changed this

File: DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs (hunk 7 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: Changed this to private

File: DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs (hunk 6 above)
Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: This is being used to get the PageSize below

Comment (kshitij-k-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Models/Common/PagingRequest.cs]: It is a private variable. As per standards it should be cameCase preceeded with `_`.
Fixed in latest PR

**Review by sameer-s-b — APPROVED** [review]

**Pull request merged**